Migrate to aiida-core 2.0

unkcpz opened the pull request:
- [x] Update readme to have a aiida-core dep matrix, since it is not backward compatible. Will then have the major version changed.

unkcpz committed:
Update to compatible with aiida-core 2.0

Following changes are made:
- deprecate 3.7 support
- Use aiida API to search and update extras field
- drop django related code

CasperWA commented on .github/workflows/ci_tests.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8, 3.9]
+        python-version: ['3.8', '3.10']
```
```suggestion
        python-version: ['3.8', '3.9', '3.10']
```
I know testing the ends should be fine, but there can always be quirks - so better to test all ranges.

CasperWA reviewed (changes requested):
You don't seem to be updating the AiiDA version in this PR?

CasperWA commented:
Ah sorry - this was still a draft - never mind me :)

unkcpz commented:
Hi @CasperWA, I see you add commits to the PR, thanks! Sorry that I am late to finish this clearly, I was distracted by other projects and just recovered from COVID. 

There are two things in the TODO list for this migration to v2.0.0: 1. There are lots of deprecation warnings need to be take care which means use the old API that not being supported. 2. For the loglevel, I change INFO to REPORT which I think it is not proper. The better solution is to use logging for aiida-optimade rather that call click log which will using aiida-core loglevel setting.

unkcpz commented:
@CasperWA I revert the echo_report to echo_info by explicitly setting aiida.cmdline log level to `INFO` for the command `init` and `calc`. The deprecation warnings are cleaned. The tests are all passed. It is ready for review.

eimrek commented:
Hi both, I tested this a bit, and if I add a structure without calling `aiida-optimade init`, then i'm getting an 500 Internal server error, see:

![image](https://user-images.githubusercontent.com/8721264/170237085-8e6af9e6-c428-46b2-bdcc-96849d05c159.png)

After calling  `aiida-optimade -p quicksetup init`, it seems to work well.

unkcpz commented:
Hi @CasperWA, do you have more expectations on what I can add to this PR? Please let me know. I think @eimrek, basically can go ahead and test based on this branch on materials cloud.

CasperWA commented:
> After calling `aiida-optimade -p quicksetup init`, it seems to work well.

At this point in time, one indeed needs to initialize/prime (run `aiida-optimade init`) before a profile can be exposed via the OPTIMADE API. So I'm not surprised. But for ease-of-use this might be a feature to implement in the future: on-the-fly initialization support.

CasperWA commented on Dockerfile:
```diff
@@ -1,10 +1,11 @@
 # Install specific optimade and aiida-core versions
-ARG OPTIMADE_TOOLS_VERSION=0.16.12
-ARG AIIDA_VERSION=1.6.8
+ARG OPTIMADE_TOOLS_VERSION=0.16.10
```
This should be updated to the latest version by rebasing on or merging in ~`master`~`develop`.

CasperWA replied:
I'll do this :)

CasperWA commented on aiida_optimade/cli/__init__.py:
```diff
@@ -1,8 +1,2 @@
-# pylint: disable=wrong-import-position
-import click_completion
```
Why is this removed? Is it not necessary anymore for completion?

sphuber replied:
Indeed, as of `click==8.0` they provide their own auto-complete which we use now.

CasperWA commented on aiida_optimade/cli/cmd_init.py:
```diff
@@ -48,6 +48,10 @@ def init(obj: dict, force: bool, silent: bool, mongo: bool, filename: str):
     from aiida import load_profile
     from aiida.cmdline.utils import echo
 
+    # The default aiida.cmdline loglevel inherit from aiida loglevel is REPORT
+    # Here we use INFO loglevel for the operations
```
Maybe copy comment to `cmd_calc.py`?

unkcpz replied:
Done.

CasperWA commented on aiida_optimade/translators/entities.py:
```diff
@@ -154,20 +154,17 @@ def _store_attributes_node_extra(self) -> None:
                 session.query(DbNode).filter(DbNode.id == self._pk).update(
                     values={"extras": extras}
                 )
+        # TODO: should be also support sqlite_zip
```
Maybe create an issue for this in this repo, if it's something that's out-of-scope for the current PR and will therefore not be implemented.
After the issue has been created, this line can be removed.

unkcpz replied:
I brought up this in one of aiida meeting. The `sqlite_zip` backend is only read-only, therefore maybe not feasible for aiida-optimade where we need to update the extra attributes for query. 

But there can be a use case that if user want to share a profile with optimade attributes already be generated, it can be used for query without actually load it. I open the issue at https://github.com/aiidateam/aiida-optimade/issues/355

CasperWA commented on requirements.txt:
```diff
@@ -1,6 +1,5 @@
-aiida-core~=1.6.8
+aiida-core[atomic_tools]~=2.0.1
```
As `atomic_tools` adds a lot of other packages, I'd prefer to have `pymatgen` be a specific, explicit package in this file and reverse this line to not include the `atomic_tools` extra.

unkcpz replied:
Yes, you are right. But I think `pymatgen` in some version break the dependency of aiida-core. If we want to make it always compatible, we need to modified the field and sure it is the same as aiida-core version is used.

I adopt your suggestion to only include `pymatgen` with version pined the same as `aiida-core` let me know if you think this is not good.

CasperWA reviewed (changes requested):
This is looking really good.

Please remove the `.vscode` folder (and maybe add it to your global `.gitignore` file, see e.g., [this StackOverflow answer](https://stackoverflow.com/a/7335487/12404091) on how to set that up).

CasperWA commented:
Let me just merge in ~`master`~`develop` properly here.

Edit: Done.

eimrek commented:
> > After calling `aiida-optimade -p quicksetup init`, it seems to work well.
> 
> At this point in time, one indeed needs to initialize/prime (run `aiida-optimade init`) before a profile can be exposed via the OPTIMADE API. So I'm not surprised. But for ease-of-use this might be a feature to implement in the future: on-the-fly initialization support.

Hi Casper, just to be clear: it's happening even when first initialization has already been run, but afterwards when adding new structures this problem occurs. E.g. it contradicts the README sentence of 

`If in the future, more StructureData nodes are added to your profile's database, these will be automatically updated for the first query, filtering on any of these OPTIMADE-specific fields`

However, i'm not sure if this problem was already there before or this PR introduced this (I currently don't have aiida < 2.0 to test quickly).

CasperWA commented:
> Hi Casper, just to be clear: it's happening even when first initialization has already been run, but afterwards when adding new structures this problem occurs. E.g. it contradicts the README sentence of
> 
> `If in the future, more StructureData nodes are added to your profile's database, these will be automatically updated for the first query, filtering on any of these OPTIMADE-specific fields`
> 
> However, i'm not sure if this problem was already there before or this PR introduced this (I currently don't have aiida < 2.0 to test quickly).

Right! Let's just create an issue for it then, and we/I can take a look at it.

Ah - just remembering now. It might be really important to inspect [this part of the code](https://github.com/aiidateam/aiida-optimade/blob/develop/aiida_optimade/translators/entities.py#L139-L178) and whether it should be updated. I needed to go directly through the specific backend to update node extras in bulk. This might have changed in the new version.

unkcpz commented on aiida_optimade/translators/entities.py:
```diff
@@ -154,20 +154,17 @@ def _store_attributes_node_extra(self) -> None:
+        if profile.storage_backend == "psql_dos":
+            # TODO: this can be `set_extra` directly? is there performance issue with sqlite_zip?
+            from aiida.storage.psql_dos.models.node import DbNode
 
             with get_manager().get_backend().transaction() as session:
                 session.query(DbNode).filter(DbNode.id == self._pk).update(
                     values={"extras": extras}
```
Here could be the issue mentioned by @eimrek rooted. As I TODO comment above, I guess you did not `load_node` and call `(re)set_extra` from performance consideration? I think there may be a way to use sqla interface to do this without performance issue? 

I'd like to @sphuber for a comment. The purpose of the code here was simply to update `extras` attributes. Is calling `load_node(<pk>)` then `set_extras` at the expense of efficiency? If so how to do it properly with sqla directly?

sphuber replied:
To be honest I am not sure anymore how much going to our front-end slows down these queries. It would be good to test since it has been refactored a lot and the slowdown may not be so bad anymore compared to before. If that is the case, I would definitely advise going through the front-end such that all storage backends are automatically supported, especially now that they are pluginnable with aiida v2.0.

CasperWA replied:
That's fair enough, and some benchmarking would be more than welcome. I'm pretty sure this will come out to still be faster, as the Node is never (supposedly) retrieved from the database, but rather a specific DB row's `extras` column is updated.

unkcpz reviewed (commented):
@CasperWA thanks for reviewing. I adopt most of your suggestions. And leave some open questions. Please have another look when you are avaliable.

unkcpz commented:
> > Hi Casper, just to be clear: it's happening even when first initialization has already been run, but afterwards when adding new structures this problem occurs. E.g. it contradicts the README sentence of
> > `If in the future, more StructureData nodes are added to your profile's database, these will be automatically updated for the first query, filtering on any of these OPTIMADE-specific fields`
> > However, i'm not sure if this problem was already there before or this PR introduced this (I currently don't have aiida < 2.0 to test quickly).
> 
> Right! Let's just create an issue for it then, and we/I can take a look at it.
> 
> Ah - just remembering now. It might be really important to inspect [this part of the code](https://github.com/aiidateam/aiida-optimade/blob/develop/aiida_optimade/translators/entities.py#L139-L178) and whether it should be updated. I needed to go directly through the specific backend to update node extras in bulk. This might have changed in the new version.

I check with old version, it also has this problem. And I guess this is why we need `aiida-optimade -p <profile> calc` to add things to extras attributes for new added `StructureData` or `CifData`? 

I open an issue from what @eimrek reported at https://github.com/aiidateam/aiida-optimade/issues/358

unkcpz commented:
I have to remove support to python3.10 for the moment, since it rely on `urllib3` and it not support python3.10 yet. I open an issue https://github.com/aiidateam/aiida-optimade/issues/359 to add it after `urllib3=2.0.0` is released or we can move to other alternative libraries.

unkcpz commented:
https://user-images.githubusercontent.com/8831685/171270334-7aa3299a-1a73-4ddd-b3cc-2a63200c220f.mov

https://user-images.githubusercontent.com/8831685/171270820-f4d65f94-a987-464a-b513-e86a09aa3b8e.mov


@sphuber @CasperWA Here are some benchmarks about direct load node and update extras with aiida API and using `DbNode`. The first record is in commit `cd2cfa3` the `DbNode` way to query and update extras. The second one is the latest commit using `node.set_extra` directly. I think there is no significant performance difference, they are all about 5-10 items/second. I think it is pretty safe to make this change.

CasperWA commented:
> Screen.Recording.2022-05-31.at.21.32.08.mov 
>  Screen.Recording.2022-05-31.at.21.35.45.mov 
> @sphuber @CasperWA Here are some benchmarks about direct load node and update extras with aiida API and using `DbNode`. The first record is in commit `cd2cfa3` the `DbNode` way to query and update extras. The second one is the latest commit using `node.set_extra` directly. I think there is no significant performance difference, they are all about 5-10 items/second. I think it is pretty safe to make this change.

It is quite difficult to determine the comparison here. Could you perhaps do these calls several times (say 5) for each method with the `time` command prepended to the `aiida-optimade` command?

Then you can plot these against each other, determining the average timings as well.

If you want to go the extra mile you could test it with different sized DBs as well.
Also, I cannot find the commits you're referencing. Maybe simply create a separate branch based off of the current one where you change to using the front-end directly?

unkcpz commented:
> It is quite difficult to determine the comparison here. Could you perhaps do these calls several times (say 5) for each method with the `time` command prepended to the `aiida-optimade` command?
> 
> Then you can plot these against each other, determining the average timings as well.
> 
> If you want to go the extra mile you could test it with different sized DBs as well. Also, I cannot find the commits you're referencing. Maybe simply create a separate branch based off of the current one where you change to using the front-end directly?

@CasperWA thanks for the advice!
I made a `cProfiler` for running init at [COF archive](https://www.materialscloud.org/discover/curated-cofs#mcloudHeader) the version that @eimrek migrate to aiida-core v2.0.1. 
From the results, I think `_store_attributes_node_extra` where I change to use aiida API didn't contribute too much walltime. But seems from the report, there are some we can improve for performance.

[optimade-profiler.txt](https://github.com/aiidateam/aiida-optimade/files/8861544/optimade-profiler.txt)

unkcpz committed:
Add support to python 3.10

unkcpz commented:
Hi @CasperWA, I think it is better to have a to-go trigger from you 😆, do you think it is good to go? 
I propose we bump the version to `v1.0.0`? what do you think?

unkcpz committed:
Merge branch 'develop' into update/aiida-v2

CasperWA commented on .github/aiida/profile.yaml:
```diff
@@ -18,3 +18,4 @@ broker_password: guest
 broker_host: 127.0.0.1
 broker_port: 5672
 broker_virtual_host: ''
+test_profile: true
```
Just adding an empty new line - not necessary, but good practice.
```suggestion
test_profile: true

```

unkcpz replied:
Yes, this will be fixed by `end-of-file-fixer`, it's been so long, no idea why I put those in excluded, seems not necessary now. 😉

CasperWA commented on .pre-commit-config.yaml:
```diff
@@ -6,6 +6,13 @@ repos:
+  - id: end-of-file-fixer
+    exclude: |
+      (?x)^(
+        .github/aiida/profile.yaml  |
```
Ah, sorry. This file _needs_ to have a non-empty EOF line? 😮

CasperWA replied:
I don't think I'd add these files here at all, but I like adding the `end-of-file-fixer` hook though, so maybe just remove the `exclude` key and value?

CasperWA commented on .pre-commit-config.yaml:
```diff
@@ -6,6 +6,13 @@ repos:
   - id: check-yaml
   - id: check-json
+  - id: end-of-file-fixer
+    exclude: |
+      (?x)^(
+        .github/aiida/profile.yaml  |
+        .github/mongo/test_structures_mongo.json |
+        .github/workflows/ci_dependabot.yml
+      )$
```
```suggestion
```

unkcpz replied:
Removed.

CasperWA reviewed (commented):
Excellent work here!

I have nothing, really, to put a finger on. Only a question regarding the `end-of-file-fixer` hook that I'd like cleared up. Then this is good to go.

Regarding updating the version, I think it's a great idea to update this to v1.0.0 for sure 👍 😄

unkcpz committed:
check all file and run end of file fixer

unkcpz committed:
Add badge for dep matrix

unkcpz commented:
Also add the dep matrix as like `aiida-quantumespresso`, can check from https://github.com/unkcpz/aiida-optimade/tree/update/aiida-v2#optimade-api-implementation-for-aiida for how it looks like

CasperWA commented on README.md:
```diff
@@ -157,3 +164,12 @@ To release the new version, go to GitHub release API of the repo create a new re
+
+[AiiDA v1.0]: https://img.shields.io/badge/AiiDA->=2.0.0,<3.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDHOtzs ...
+
+[AiiDA v0.18]: https://img.shields.io/badge/AiiDA->=1.6.0,<2.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDHOtz ...
```
```suggestion
[AiiDA v1 range]: https://img.shields.io/badge/AiiDA->=1.0.0,<2.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDHOtzsz5z%2B%2FuZl7ztmF%2F5HJvxVQN6cPYX8%2FPLnOmsvNAvqfwuib%2FbNIk9cQeQnLcKRL5xLIV%2Fic9eJeunjPYbRs4FjQSpTB3aS1IpRKeeOOewajy%2FKKEO8Q0DuVdKy8IqsbPulxGHUfCBBu%2BwUYGuFuBTK7wQnht6PEbf4tlRomVRjCbXNjQEB0AyrFQOL5ENIJm7dTLZE6DPJCnEtFZVXDLny%2B4Sjv0PmmYu1ZdUek9RiMgoDmJ8V0L7XJqsZ3UW8YsBOwEeHeeFce7jEYXBy0m9m4BbXqSj2%2Bxnkg26MCVrN6DEZcwggtd8pTFx%2Fh3B9B50YLaFOPwXQKUt0tBLegtSomfBlfY13PwijbEnhztGzgJsK5h9W9qeWwBqjvyhB2iBs1Qz0AU974DciRGO8CVN8AJhAeMAdA3KbrKEtvxhsI%2B9emWiJlGBEU680Cfk%2BSsVqXZvcFYGXjF8ABVJ%2BTNfVXehyms1zzn1gmIOxLEB6E31%2FWBe5rnCarmo7elf7dJEeaLh80GasliI5F6Q9cAz1GY1OJVNDxTzQTw7iY%2FHEZRQY7xqJ9RU2LFe%2FYqakdP911ha0XhjjiTVAkDwgatWfCGeYocx8M3glG8g8EXhSrLrHnEFJ5Ymow%2FkhIYv6ttYUW1iFmEqqxdVoUs9FmsDYSqmtmJh3Cl1%2BVtl2s7owDUdocR5bceiyoSivGTT5vzpbzL1uoBpmcAAQgW7ArnKD9ng9rc%2BNgrobSNwpSkkhcRN%2BvmXLjIsDovYHHEfmsYFygPAnIDEQrQPzJYCOaLHLUfIt7Oq0LJn9fxkSgNCb1qEIQ5UKgT%2Fs6gJmVOOroJhQBXVqw118QtWLdyUxEP45sUpSzqP7RDdFYMyB9UReMiF1MzPwoUqHt8hjGFFeP5wZAbZ%2F0%2BcAtAAcji6LeSq%2FMYiAvSsdw3GtrfVSVFUBbIhwRWYR7yOcr%2FBi%2FB1MSJZ16JlgH1AGM3EO2QnmMyrSbTSiACgFBv4yCUapZkt9qwWVL7aeOyHvArJjm8%2Fz9BhdI4XcZgz2%2FvRALosjsk1ODOyMcJn9%2FYI6IrkS5vxMGdUwou2YKfyVqJpn5t9aNs3gbQMbdbkxnGdsr4bTHm2AxWo9yNZK4PXR3uzhAh%2BM0AZejnCrGdy0UvJxl0oMKgWSLR%2B1LH2aE9ViejiFs%2BXn6bTjng3MlIhJ1I1TkuLdg6OcAbD7Xx%2Bc3y9TrWAiSHqVkbZ2v9ilCo6s4AjwZCzFyD9mOL305nV9aonvsQeT2L0gVk4OwOJqXXVRW7naaxswDKVdlYLyMXAnntteYmws2xcVVZzq%2BtHPAooQggmJkc6TLSusOiL4RKgwzzYU1iFQgiUBA1H7E8yPau%2BZl9P7AblVNebtHqTgxLfRqrNvZWjsHZFuqMqKcDWdlFjF7UGvX8Jn24DyEAykJwNcdg0OvJ4p5pQ9tV6SMlP4A0PNh8aYze1ArROyUNTNouy8tNF3Rt0CSXb6bRFl4%2FIfQzNMjaE9WwpYOWQnOdEF%2BTdJNO0iFh7%2BI0kfORzQZb6P2kymS9oTxzBiM9rUqLWr1WE5G6ODhycQd%2FUnNVeMbcH68hYkGycNoUNWc8fxaxfwhDbHpfwM5oeTY7rUX8QAAAABJRU5ErkJggg%3D%3D
```

CasperWA commented on README.md:
```diff
@@ -157,3 +164,12 @@ To release the new version, go to GitHub release API of the repo create a new re
 Note, the tag should start with a `v` and be followed by a full semantic version (see [SemVer](https://semver.org)).
 For example: `v2.3.12`.
+
+[AiiDA v1.0]: https://img.shields.io/badge/AiiDA->=2.0.0,<3.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDHOtzs ...
```
```suggestion
[AiiDA v2 range]: https://img.shields.io/badge/AiiDA->=2.0.0,<3.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDHOtzsz5z%2B%2FuZl7ztmF%2F5HJvxVQN6cPYX8%2FPLnOmsvNAvqfwuib%2FbNIk9cQeQnLcKRL5xLIV%2Fic9eJeunjPYbRs4FjQSpTB3aS1IpRKeeOOewajy%2FKKEO8Q0DuVdKy8IqsbPulxGHUfCBBu%2BwUYGuFuBTK7wQnht6PEbf4tlRomVRjCbXNjQEB0AyrFQOL5ENIJm7dTLZE6DPJCnEtFZVXDLny%2B4Sjv0PmmYu1ZdUek9RiMgoDmJ8V0L7XJqsZ3UW8YsBOwEeHeeFce7jEYXBy0m9m4BbXqSj2%2Bxnkg26MCVrN6DEZcwggtd8pTFx%2Fh3B9B50YLaFOPwXQKUt0tBLegtSomfBlfY13PwijbEnhztGzgJsK5h9W9qeWwBqjvyhB2iBs1Qz0AU974DciRGO8CVN8AJhAeMAdA3KbrKEtvxhsI%2B9emWiJlGBEU680Cfk%2BSsVqXZvcFYGXjF8ABVJ%2BTNfVXehyms1zzn1gmIOxLEB6E31%2FWBe5rnCarmo7elf7dJEeaLh80GasliI5F6Q9cAz1GY1OJVNDxTzQTw7iY%2FHEZRQY7xqJ9RU2LFe%2FYqakdP911ha0XhjjiTVAkDwgatWfCGeYocx8M3glG8g8EXhSrLrHnEFJ5Ymow%2FkhIYv6ttYUW1iFmEqqxdVoUs9FmsDYSqmtmJh3Cl1%2BVtl2s7owDUdocR5bceiyoSivGTT5vzpbzL1uoBpmcAAQgW7ArnKD9ng9rc%2BNgrobSNwpSkkhcRN%2BvmXLjIsDovYHHEfmsYFygPAnIDEQrQPzJYCOaLHLUfIt7Oq0LJn9fxkSgNCb1qEIQ5UKgT%2Fs6gJmVOOroJhQBXVqw118QtWLdyUxEP45sUpSzqP7RDdFYMyB9UReMiF1MzPwoUqHt8hjGFFeP5wZAbZ%2F0%2BcAtAAcji6LeSq%2FMYiAvSsdw3GtrfVSVFUBbIhwRWYR7yOcr%2FBi%2FB1MSJZ16JlgH1AGM3EO2QnmMyrSbTSiACgFBv4yCUapZkt9qwWVL7aeOyHvArJjm8%2Fz9BhdI4XcZgz2%2FvRALosjsk1ODOyMcJn9%2FYI6IrkS5vxMGdUwou2YKfyVqJpn5t9aNs3gbQMbdbkxnGdsr4bTHm2AxWo9yNZK4PXR3uzhAh%2BM0AZejnCrGdy0UvJxl0oMKgWSLR%2B1LH2aE9ViejiFs%2BXn6bTjng3MlIhJ1I1TkuLdg6OcAbD7Xx%2Bc3y9TrWAiSHqVkbZ2v9ilCo6s4AjwZCzFyD9mOL305nV9aonvsQeT2L0gVk4OwOJqXXVRW7naaxswDKVdlYLyMXAnntteYmws2xcVVZzq%2BtHPAooQggmJkc6TLSusOiL4RKgwzzYU1iFQgiUBA1H7E8yPau%2BZl9P7AblVNebtHqTgxLfRqrNvZWjsHZFuqMqKcDWdlFjF7UGvX8Jn24DyEAykJwNcdg0OvJ4p5pQ9tV6SMlP4A0PNh8aYze1ArROyUNTNouy8tNF3Rt0CSXb6bRFl4%2FIfQzNMjaE9WwpYOWQnOdEF%2BTdJNO0iFh7%2BI0kfORzQZb6P2kymS9oTxzBiM9rUqLWr1WE5G6ODhycQd%2FUnNVeMbcH68hYkGycNoUNWc8fxaxfwhDbHpfwM5oeTY7rUX8QAAAABJRU5ErkJggg%3D%3D
```

CasperWA commented on README.md:
```diff
@@ -1,5 +1,12 @@
+
+| Plugin | AiiDA | Python | Specification |
+|-|-|-|-|
+| `v1.0 < v2.0` | ![Compatibility for v1.0][AiiDA v1.0] |  [![PyPI pyversions](https://img.shields.io/pypi/pyversions/aiida-optimade)](https://pypi.org/project/aiida-optimade) | ![OPTIMADE SPEC compatibility][OPTIMADE spec v0.18] |
```
```suggestion
| `v1.0 < v2.0` | ![Compatibility for v1][AiiDA v2 range] |  [![PyPI pyversions](https://img.shields.io/pypi/pyversions/aiida-optimade)](https://pypi.org/project/aiida-optimade) | ![OPTIMADE API compatibility][OPTIMADE from OPT] |
```

CasperWA commented on README.md:
```diff
@@ -157,3 +164,12 @@ To release the new version, go to GitHub release API of the repo create a new re
+[Python v3.7-v3.9]: https://img.shields.io/badge/python-3.7%20%7C%203.8%20%7C%203.9-blue
+[Python v3.8-v3.10]: https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C%203.10-blue
+
+[OPTIMADE spec v0.18]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/Materials-Consortia/optimade-python-tools/v0.18.0/optimade-version.json
```
```suggestion
[OPTIMADE from OPT]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/Materials-Consortia/optimade-python-tools/v0.18.0/optimade-version.json
```

CasperWA commented on README.md:
```diff
@@ -1,5 +1,12 @@
+| Plugin | AiiDA | Python | Specification |
+|-|-|-|-|
+| `v1.0 < v2.0` | ![Compatibility for v1.0][AiiDA v1.0] |  [![PyPI pyversions](https://img.shields.io/pypi/pyversions/aiida-optimade)](https://pypi.org/project/aiida-optimade) | ![OPTIMADE SPEC compatibility][OPTIMADE spec v0.18] |
+| `v0.18 < v0.20` | ![Compatibility for v3.5][AiiDA v0.18] |  [![PyPI pyversions][Python v3.7-v3.9]](https://pypi.org/project/aiida-optimade/0.18.0/) | ![OPTIMADE SPEC compatibility][OPTIMADE spec v0.18] |
```
```suggestion
| `v0.18 <= v0.20` | ![Compatibility for v0][AiiDA v1 range] |  [![PyPI pyversions][Python v3.7-v3.9]](https://pypi.org/project/aiida-optimade/0.20.0/) | ![OPTIMADE API compatibility][OPTIMADE from OPT] |
```

CasperWA reviewed (changes requested):
The matrix looks nice - although it's a bit confusing with the naming, so I changed that a bit. Also, there is no record of this repository pre v0.18, which is a bit confusing as well, but I suppose it's not that relevant.

unkcpz committed:
review dep matx

unkcpz commented:
@CasperWA thanks! All changed.

CasperWA commented on README.md:
```diff
@@ -157,3 +164,12 @@ To release the new version, go to GitHub release API of the repo create a new re
+[AiiDA v1 range]: https://img.shields.io/badge/AiiDA->=1.6.0,<2.0.0-007ec6.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAhCAYAAABTERJSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAFhgAABYYBG6Yz4AAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAUbSURBVFiFzZhrbFRVEMd%2Fc%2B5uu6UUbIFC%2FUAUVEQCLbQJBIiBDyiImJiIhmohYNCkqJAQxASLF8tDgYRHBLXRhIcKNtFEhVDgAxBJqgmVh4JEKg3EIn2QYqBlt917xg%2BFss%2ByaDH ...
+
+[Python v3.7-v3.9]: https://img.shields.io/badge/python-3.7%20%7C%203.8%20%7C%203.9-blue
+[Python v3.8-v3.10]: https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C%203.10-blue
```
Not used
```suggestion
```

unkcpz replied:
Thanks! changed.

unkcpz committed:
r

CasperWA reviewed (dismissed):
Great work @unkcpz !

CasperWA commented:
I'll add a fix for the CI to use `optimade` version 0.18 instead of the new 0.19.

CasperWA committed:
Use OPTIMADE validator with v0.18.0

This is temporary until aiida-optimade works with OPTIMADE Python tools
v0.19 and above.

unkcpz commented:
@CasperWA thanks! But seems the change fail the tests, any idea? I can have a close look into it later today, or do you already have a clue and working on it?

CasperWA commented:
> @CasperWA thanks! But seems the change fail the tests, any idea? I can have a close look into it later today, or do you already have a clue and working on it?

Yeah - I have no idea why, `response_fields` should be a valid query parameter... And at least it should work with the older version.

CasperWA commented:
It seems OPTIMADE Python tools has the [same issue](https://github.com/Materials-Consortia/optimade-python-tools/runs/8137260571?check_suite_focus=true#step:6:165)

Indeed, a work-around is underway, see Materials-Consortia/optimade-python-tools#1317.

ml-evs commented:
> It seems OPTIMADE Python tools has the [same issue](https://github.com/Materials-Consortia/optimade-python-tools/runs/8137260571?check_suite_focus=true#step:6:165)
> 
> Indeed, a work-around is underway, see [Materials-Consortia/optimade-python-tools#1317](https://github.com/Materials-Consortia/optimade-python-tools/pull/1317).

New pydantic release is due in a few minutes, so the old Dockerfile will work again. I also pinned the Docker deps in https://github.com/Materials-Consortia/optimade-python-tools/pull/1315 so this problem shouldn't occur again.

unkcpz commented:
@ml-evs thanks a lot! I retrigger the tests, hope it works fine now.

ml-evs commented:
I just released [v0.19.2](https://github.com/Materials-Consortia/optimade-python-tools/releases/tag/v0.19.2) which includes the fix for self-built docker images (which I think is what you are doing here) and also the 2D structure CIF export bug that was reported at the workshop. I'd recommend upgrading!

unkcpz commented:
@ml-evs, thanks! I will check the update to `0.19.2` which seems break something as show in https://github.com/aiidateam/aiida-optimade/pull/375
However, now as here @CasperWA sticking the validator version to `0.18.0` and it should work, no?

ml-evs commented:
> @ml-evs, thanks! I will check the update to `0.19.2` which seems break something as show in #375 However, now as here @CasperWA sticking the validator version to `0.18.0` and it should work, no?

The 3.7 test failures are caused by us dropping Python 3.7 support, which we did mostly because AiiDA 2.0 also dropped it 😅 The other tests look like you just need to pass through the config value for `meta.schemas` in some of your custom router methods. I can patch this if you like (see https://github.com/Materials-Consortia/optimade-python-tools/pull/1210); though I guess you may want to do it outside of this PR!

unkcpz commented:
I don't understand why this is related to py3.7, the CI docker test is all run for py3.8. 
Please feel free to add a patch to fix the CI test here, thanks a lot!

ml-evs commented:
> I don't understand why this is related to py3.7, the CI docker test is all run for py3.8. Please feel free to add a patch to fix the CI test here, thanks a lot!

I only meant for the other PR you linked, #375. Here, everything should be fixed if you change `validator_version` to `latest` (or 0.19) rather than 0.18.0

unkcpz commented:
I see, thank you! I'll take care of that.

CasperWA committed:
Revert "Use OPTIMADE validator with v0.18.0"

This reverts commit dd1b909c036f264811c3ff3e5dfe1a31086a217f.

CasperWA commented:
> I see, thank you! I'll take care of that.

I just reverted my latest "fix" then - I do indeed think we need to add the `schema` parameter in some places - however, that should really coincide with the update to OPTIMADE Python tools v0.19 (i.e., #375). Also, we should definitely use the new Docker image created in the OPTIMATE Python tools repository (see [here](https://github.com/Materials-Consortia/optimade-python-tools/pkgs/container/optimade)).
Although the latter may not be possible due to the extra Docker configurations.

unkcpz commented:
> Also, we should definitely use the new Docker image created in the OPTIMATE Python tools repository (see [here](https://github.com/Materials-Consortia/optimade-python-tools/pkgs/container/optimade)).

I agree. I open an issue at https://github.com/aiidateam/aiida-optimade/issues/379. Not sure if all the image parameters are configurable. I have used it to replace the dedicated branch in [materials cloud for optimade-index](https://github.com/materialscloud-org/openstack-ansible/pull/158)
